src/actors/builtin/multisig/multisig_actor.go

```diff
@@ -18,20 +18,98 @@ var IMPL_FINISH = autil.IMPL_FINISH
 var IMPL_TODO = autil.IMPL_TODO
 var TODO = autil.TODO
 
-////////////////////////////////////////////////////////////////////////////////
-// Actor methods
-////////////////////////////////////////////////////////////////////////////////
+type TxnID int
 
-func (a *MultiSigActorCode_I) Propose(rt vmr.Runtime, txn MultiSigTransaction) TxnID {
+type MultiSigTransaction struct {
+	Proposer   addr.Address
+	Expiration abi.ChainEpoch
+
+	To     addr.Address
+	Method abi.MethodNum
+	Params abi.MethodParams
+	Value  abi.TokenAmount
+}
+
+func (txn *MultiSigTransaction) Equals(MultiSigTransaction) bool {
+	IMPL_FINISH()
+	panic("")
+}
+
+type MultiSigTransactionHAMT map[TxnID]MultiSigTransaction
+type MultiSigApprovalSetHAMT map[TxnID]autil.ActorIDSetHAMT
+
+func MultiSigTransactionHAMT_Empty() MultiSigTransactionHAMT {
+	IMPL_FINISH()
+	panic("")
+}
+
+func MultiSigApprovalSetHAMT_Empty() MultiSigApprovalSetHAMT {
+	IMPL_FINISH()
+	panic("")
+}
+
+type MultiSigActorState struct {
+	// Linear unlock
+	InitialBalance abi.TokenAmount
+	StartEpoch     abi.ChainEpoch
+	UnlockDuration abi.ChainEpoch
+
+	AuthorizedParties     autil.ActorIDSetHAMT
+	NumApprovalsThreshold int
+	NextTxnID             TxnID
+	PendingTxns           MultiSigTransactionHAMT
+	PendingApprovals      MultiSigApprovalSetHAMT
+}
+
+func (st *MultiSigActorState) AmountLocked(elapsedEpoch abi.ChainEpoch) abi.TokenAmount {
+	if elapsedEpoch >= st.UnlockDuration {
+		return abi.TokenAmount(0)
+	}
+
+	TODO() // BigInt
+	lockedProportion := (st.UnlockDuration - elapsedEpoch) / st.UnlockDuration
+	return abi.TokenAmount(uint64(st.InitialBalance) * uint64(lockedProportion))
+}
+
+// return true if MultiSig maintains required locked balance after spending the amount
+func (st *MultiSigActorState) _hasAvailable(currBalance abi.TokenAmount, amountToSpend abi.TokenAmount, currEpoch abi.ChainEpoch) bool {
+	if amountToSpend < 0 || currBalance < amountToSpend {
+		return false
+	}
+
+	if currBalance-amountToSpend < st.AmountLocked(currEpoch-st.StartEpoch) {
+		return false
+	}
+
+	return true
+}
+
+func (st *MultiSigActorState) CID() cid.Cid {
+	panic("TODO")
+}
```

Comment:
Rather than move it up here, le'ts split this to a _state.go file like the other actors.